hoodie-client/src/main/java/com/uber/hoodie/io/HoodieAppendHandle.java

```diff
@@ -162,35 +169,59 @@ private void init(Iterator<HoodieRecord<T>> recordItr) {
 
   public void doAppend() {
 
-    List<IndexedRecord> recordList = new ArrayList<>();
-    List<String> keysToDelete = new ArrayList<>();
+    int maxBlockSize = config.getLogFileDataBlockMaxSize();
+    int maxAllowedRecords = (int) (maxBlockSize / averageRecordSize);
     Map<HoodieLogBlock.LogMetadataType, String> metadata = Maps.newHashMap();
     metadata.put(HoodieLogBlock.LogMetadataType.INSTANT_TIME, commitTime);
-    Iterator<HoodieRecord<T>> recordsItr = records.iterator();
-    while (recordsItr.hasNext()) {
-      HoodieRecord record = recordsItr.next();
-      Optional<IndexedRecord> indexedRecord = getIndexedRecord(record);
-      if (indexedRecord.isPresent()) {
-        recordList.add(indexedRecord.get());
-      } else {
-        keysToDelete.add(record.getRecordKey());
+    while (recordItr.hasNext()) {
+      HoodieRecord record = recordItr.next();
+      if(!addAndCheckIfFlushNeeded(record, maxAllowedRecords)) {
+        continue;
       }
-      recordsItr.remove(); //remove entries when IndexedRecord added to new list
+      doAppend(metadata);
+      maxAllowedRecords = (int) (maxBlockSize / averageRecordSize);
     }
+    doAppend(metadata);
+  }
+
+  private void doAppend(Map<HoodieLogBlock.LogMetadataType, String> metadata) {
     try {
       if (recordList.size() > 0) {
         writer = writer.appendBlock(new HoodieAvroDataBlock(recordList, schema, metadata));
+        recordList.clear();
       }
       if (keysToDelete.size() > 0) {
         writer = writer.appendBlock(
             new HoodieDeleteBlock(keysToDelete.stream().toArray(String[]::new), metadata));
+        keysToDelete.clear();
       }
     } catch (Exception e) {
       throw new HoodieAppendException(
           "Failed while appending records to " + currentLogFile.getPath(), e);
     }
   }
 
+  /**
+   * Add record to list, check and flush if max number of records reached to achieve block size
+   * @param record
+   * @param maxAllowedRecords
+   * @return
+   */
+  private boolean addAndCheckIfFlushNeeded(HoodieRecord record, int maxAllowedRecords) {
+    Optional<IndexedRecord> indexedRecord = getIndexedRecord(record);
+    if (indexedRecord.isPresent()) {
+      recordList.add(indexedRecord.get());
+    } else {
+      keysToDelete.add(record.getRecordKey());
+    }
+    if(recordList.size() + keysToDelete.size() >= maxAllowedRecords) {
+      // Recompute averageRecordSize and update existing value with avg of new and old
+      averageRecordSize = (averageRecordSize + SizeEstimator.estimate(record))/2;
```

Comment:
its okay to call this in a loop?  omkar, had some concerns about this previously..

Reply:
This get's called only at the end of each block appended.